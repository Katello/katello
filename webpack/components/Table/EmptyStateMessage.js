import React from 'react';
import { EmptyState,
  EmptyStateBody,
  EmptyStateIcon,
  EmptyStateVariant,
  Bullseye,
  Title } from '@patternfly/react-core';
import PropTypes from 'prop-types';
import { CubeIcon, ExclamationCircleIcon, SearchIcon } from '@patternfly/react-icons';
import { global_danger_color_200 as dangerColor } from '@patternfly/react-tokens';

const KatelloEmptyStateIcon = ({ error, search }) => {
  if (error) return <EmptyStateIcon icon={ExclamationCircleIcon} color={dangerColor.value} />;
  if (search) return <EmptyStateIcon icon={SearchIcon} />;
  return <EmptyStateIcon icon={CubeIcon} />;
};

const EmptyStateMessage = ({
  title, body, error, search,
}) => (
  <Bullseye>
    <EmptyState variant={EmptyStateVariant.small}>
      <KatelloEmptyStateIcon error={!!error} search={search} />
      <Title headingLevel="h2" size="lg">
        {title}
      </Title>
      <EmptyStateBody>
        {body}
      </EmptyStateBody>
    </EmptyState>
  </Bullseye>
);

KatelloEmptyStateIcon.propTypes = {
  error: PropTypes.bool,
  search: PropTypes.bool,
};

KatelloEmptyStateIcon.defaultProps = {
  error: false,
  search: false,
};

EmptyStateMessage.propTypes = {
  title: PropTypes.string,
  body: PropTypes.string,
  error: PropTypes.oneOfType([
    PropTypes.shape({}),
    PropTypes.string,
  ]),
  search: PropTypes.bool,
};

EmptyStateMessage.defaultProps = {
  title: 'Unable to connect',
  body: 'There was an error retrieving data from the server. Check your connection and try again.',
  error: undefined,
  search: false,
};

export default EmptyStateMessage;
