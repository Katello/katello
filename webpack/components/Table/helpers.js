// Can be included as a TableWrapper prop for selectable rows
const onSelect = (rows, setRows) => (_event, isSelected, rowId) => {
  let newRows;
  if (rowId === -1) {
    newRows = rows.map(row => ({ ...row, selected: isSelected }));
  } else {
    newRows = [...rows];
    newRows[rowId].selected = isSelected;
  }

  setRows(newRows);
};

export default onSelect;
