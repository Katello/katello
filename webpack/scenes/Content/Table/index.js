import ContentTable from './ContentTable';

export default ContentTable;
