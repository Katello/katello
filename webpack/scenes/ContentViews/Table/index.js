import ContentViewTable from './ContentViewTable';

export default ContentViewTable;
