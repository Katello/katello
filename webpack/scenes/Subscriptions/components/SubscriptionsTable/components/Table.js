import React from 'react';
import PropTypes from 'prop-types';
import { Table as PFtable } from 'patternfly-react';
import classNames from 'classnames';
import { createSubscriptionsTableSchema } from '../SubscriptionsTableSchema';
import { Table as ForemanTable, TableBody as ForemanTableBody } from '../../../../../move_to_foreman/components/common/table';

const Table = ({
  emptyState,
  tableColumns,
  subscriptions,
  loadSubscriptions,
  selectionController,
  inlineEditController,
  rows,
  editing,
  groupedSubscriptions,
}) => {
  const allSubscriptionResults = subscriptions.results;

  let bodyMessage;
  if (allSubscriptionResults.length === 0 && subscriptions.searchIsActive) {
    bodyMessage = __('No subscriptions match your search criteria.');
  }

  const groupingController = {
    isCollapseable: ({ rowData }) =>
      // it is the first subscription in the group
      rowData.id === groupedSubscriptions[rowData.product_id].subscriptions[0].id &&
      // the group contains more then one subscription
      groupedSubscriptions[rowData.product_id].subscriptions.length > 1,
    isCollapsed: ({ rowData }) => !groupedSubscriptions[rowData.product_id].open,
    toggle: ({ rowData }) => this.toggleSubscriptionGroup(rowData.product_id),
  };

  const alwaysDisplayColumns = ['select'];
  const columnsDefinition = createSubscriptionsTableSchema(
    inlineEditController,
    selectionController,
    groupingController,
  ).filter(column => tableColumns.includes(column.property) ||
    alwaysDisplayColumns.includes(column.property));

  const onPaginationChange = (pagination) => {
    loadSubscriptions({
      ...pagination,
    });
  };


  return (
    <ForemanTable
      columns={columnsDefinition}
      emptyState={emptyState}
      bodyMessage={bodyMessage}
      rows={rows}
      components={{
      header: {
        row: PFtable.TableInlineEditHeaderRow,
      },
    }}
      itemCount={subscriptions.itemCount}
      pagination={subscriptions.pagination}
      onPaginationChange={onPaginationChange}
      inlineEdit
    >
      <PFtable.Header
        onRow={() => ({
        role: 'row',
        isEditing: () => editing,
        onCancel: () => inlineEditController.onCancel(),
        onConfirm: () => inlineEditController.onConfirm(),
      })}
      />
      <ForemanTableBody
        columns={columnsDefinition}
        rows={rows}
        rowKey="id"
        message={bodyMessage}
        onRow={rowData => ({
        className: classNames({ 'open-grouped-row': !groupingController.isCollapsed({ rowData }) }),
      })}
      />
    </ForemanTable>
  );
};

Table.propTypes = {
  emptyState: PropTypes.shape({}).isRequired,
  tableColumns: PropTypes.arrayOf(PropTypes.string).isRequired,
  subscriptions: PropTypes.shape({
    searchIsActive: PropTypes.bool,
    itemCount: PropTypes.number,
    pagination: PropTypes.shape({}),
    results: PropTypes.array,
  }).isRequired,
  loadSubscriptions: PropTypes.func.isRequired,
  selectionController: PropTypes.shape({}).isRequired,
  inlineEditController: PropTypes.shape({
    onCancel: PropTypes.func,
    onConfirm: PropTypes.func,
  }).isRequired,
  groupedSubscriptions: PropTypes.shape({}).isRequired,
  editing: PropTypes.bool.isRequired,
  rows: PropTypes.arrayOf(PropTypes.object).isRequired,

};

export default Table;
