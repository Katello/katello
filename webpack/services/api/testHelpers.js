
export const failureAction = (type, message = 'Request failed with status code 422') => (
  {
    type,
    payload: {
      messages: [message],
      result: new Error(message),
    },
  }
);

export const toastErrorAction = (message = 'Request failed with status code 422') => (
  {
    payload: {
      message: {
        message,
        sticky: true,
        type: 'error',
      },
    },
    type: 'TOASTS_ADD',
  }
);

export default {
  failureAction,
  toastErrorAction,
};
