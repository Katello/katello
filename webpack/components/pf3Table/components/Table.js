import React from 'react';
import PropTypes from 'prop-types';
import { Table as PfTable } from 'patternfly-react';
import { noop } from 'foremanReact/common/helpers';
import EmptyState from 'foremanReact/components/common/EmptyState';
import Pagination from 'foremanReact/components/Pagination';

import TableBody from './TableBody';

const Table = ({
  columns,
  rows,
  emptyState,
  bodyMessage,
  children,
  itemCount,
  pagination,
  onPaginationChange,
  ...props
}) => {
  if (rows.length === 0 && bodyMessage === undefined) {
    return <EmptyState {...emptyState} />;
  }

  const shouldRenderPagination = itemCount && pagination;

  const body = children || [
    <PfTable.Header key="header" />,
    <TableBody key="body" columns={columns} rows={rows} message={bodyMessage} rowKey="id" />,
  ];

  return (
    <div>
      <PfTable.PfProvider
        columns={columns}
        className="table-fixed neat-table-cells"
        striped
        bordered
        hover
        {...props}
      >
        {body}
      </PfTable.PfProvider>
      {shouldRenderPagination && (
        <Pagination
          itemCount={itemCount}
          onChange={onPaginationChange}
          {...pagination}
        />
      )}
    </div>
  );
};

Table.propTypes = {
  columns: PropTypes.arrayOf(PropTypes.shape({})).isRequired,
  rows: PropTypes.arrayOf(PropTypes.shape({})).isRequired,
  emptyState: PropTypes.object, // eslint-disable-line react/forbid-prop-types
  pagination: PropTypes.object, // eslint-disable-line react/forbid-prop-types
  bodyMessage: PropTypes.node,
  children: PropTypes.node,
  itemCount: PropTypes.number,
  onPaginationChange: PropTypes.func,
};

Table.defaultProps = {
  emptyState: undefined,
  pagination: undefined,
  bodyMessage: undefined,
  children: undefined,
  itemCount: undefined,
  onPaginationChange: noop,
};

export default Table;
