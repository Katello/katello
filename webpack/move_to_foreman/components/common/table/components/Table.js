import React from 'react';
import PropTypes from 'prop-types';
import { Table as PfTable } from 'patternfly-react';
import { noop } from 'foremanReact/common/helpers';
import EmptyState from 'foremanReact/components/common/EmptyState';
import Pagination from 'foremanReact/components/Pagination/PaginationWrapper';

import TableBody from './TableBody';

const Table = ({
  columns,
  rows,
  emptyState,
  bodyMessage,
  children,
  itemCount,
  pagination,
  onPaginationChange,
  rowKey,
  ...props
}) => {
  if (rows.length === 0 && bodyMessage === undefined) {
    return <EmptyState {...emptyState} />;
  }

  const shouldRenderPagination = itemCount && pagination;

  const body = children || [
    <PfTable.Header key="header" />,
    <TableBody key="body" columns={columns} rows={rows} message={bodyMessage} rowKey={rowKey} />,
  ];

  return (
    <div>
      <PfTable.PfProvider
        columns={columns}
        className="table-fixed"
        striped
        bordered
        hover
        {...props}
      >
        {body}
      </PfTable.PfProvider>
      {shouldRenderPagination && (
        <Pagination
          viewType="table"
          itemCount={itemCount}
          pagination={pagination}
          onChange={onPaginationChange}
        />
      )}
    </div>
  );
};

Table.propTypes = {
  columns: PropTypes.arrayOf(PropTypes.object).isRequired,
  rows: PropTypes.arrayOf(PropTypes.object).isRequired,
  emptyState: PropTypes.object, // eslint-disable-line react/forbid-prop-types
  pagination: PropTypes.object, // eslint-disable-line react/forbid-prop-types
  bodyMessage: PropTypes.node,
  children: PropTypes.node,
  itemCount: PropTypes.number,
  onPaginationChange: PropTypes.func,
  rowKey: PropTypes.string,
};

Table.defaultProps = {
  emptyState: undefined,
  pagination: undefined,
  bodyMessage: undefined,
  children: undefined,
  itemCount: undefined,
  rowKey: 'id',
  onPaginationChange: noop,
};

export default Table;
