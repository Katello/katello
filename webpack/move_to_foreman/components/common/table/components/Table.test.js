import { testComponentSnapshotsWithFixtures } from '../../../../../move_to_pf/test-utils/testHelpers';

import Table from './Table';
import { columnsFixtures, rowsFixtures } from './TableFixtures';

const fixtures = {
  'renders Table with emptyState': {
    columns: columnsFixtures,
    rows: [],
    emptyState: { empty: 'state' },
  },
  'renders Table with children': {
    columns: columnsFixtures,
    rows: rowsFixtures,
    children: 'some children',
  },
  'renders Table with body': {
    columns: columnsFixtures,
    rows: rowsFixtures,
    bodyMessage: 'some body message',
  },
  'renders Table with pagination': {
    columns: columnsFixtures,
    rows: rowsFixtures,
    children: 'some children',
    itemCount: 2,
    pagination: { page: 1, perPage: 20 },
  },
};

describe('Table', () => testComponentSnapshotsWithFixtures(Table, fixtures));
