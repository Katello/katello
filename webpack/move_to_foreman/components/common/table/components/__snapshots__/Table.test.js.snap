// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table renders Table with body 1`] = `
<div>
  <TablePfProvider
    bordered={true}
    className="table-fixed"
    columns={
      Array [
        Object {
          "header": Object {
            "label": "ID",
          },
          "property": "id",
        },
        Object {
          "header": Object {
            "label": "Data",
          },
          "property": "data",
        },
      ]
    }
    components={Object {}}
    condensed={false}
    dataTable={false}
    hover={true}
    inlineEdit={false}
    striped={true}
  >
    <Header
      key="header"
    />
    <TableBody
      columns={
        Array [
          Object {
            "header": Object {
              "label": "ID",
            },
            "property": "id",
          },
          Object {
            "header": Object {
              "label": "Data",
            },
            "property": "data",
          },
        ]
      }
      key="body"
      message="some body message"
      rowKey="id"
      rows={
        Array [
          Object {
            "data": "data-1",
            "id": 1,
          },
          Object {
            "data": "data-2",
            "id": 2,
          },
        ]
      }
    />
  </TablePfProvider>
</div>
`;

exports[`Table renders Table with children 1`] = `
<div>
  <TablePfProvider
    bordered={true}
    className="table-fixed"
    columns={
      Array [
        Object {
          "header": Object {
            "label": "ID",
          },
          "property": "id",
        },
        Object {
          "header": Object {
            "label": "Data",
          },
          "property": "data",
        },
      ]
    }
    components={Object {}}
    condensed={false}
    dataTable={false}
    hover={true}
    inlineEdit={false}
    striped={true}
  >
    some children
  </TablePfProvider>
</div>
`;

exports[`Table renders Table with emptyState 1`] = `
<EmptyState
  empty="state"
/>
`;

exports[`Table renders Table with pagination 1`] = `
<div>
  <TablePfProvider
    bordered={true}
    className="table-fixed"
    columns={
      Array [
        Object {
          "header": Object {
            "label": "ID",
          },
          "property": "id",
        },
        Object {
          "header": Object {
            "label": "Data",
          },
          "property": "data",
        },
      ]
    }
    components={Object {}}
    condensed={false}
    dataTable={false}
    hover={true}
    inlineEdit={false}
    striped={true}
  >
    some children
  </TablePfProvider>
  <PaginationRow
    className=""
    dropdownButtonId="pagination-row-dropdown"
    itemCount={2}
    messages={
      Object {
        "currentPage": "Current Page",
        "firstPage": "First Page",
        "lastPage": "Last Page",
        "nextPage": "Next Page",
        "of": "of",
        "perPage": "per page",
        "previousPage": "Previous Page",
      }
    }
    onChange={[Function]}
    onPageSet={[Function]}
    onPerPageSelect={[Function]}
    pagination={
      Object {
        "page": 1,
        "perPage": 20,
      }
    }
    viewType="table"
  />
</div>
`;
